core/src/Kokkos_MathematicalFunctions.hpp

```diff
@@ -55,124 +55,233 @@
 #endif
 
 namespace Kokkos {
+
+namespace Impl {
+template <class T, bool = std::is_integral<T>::value>
+struct promote {
+  using type = double;
+};
+template <class T>
+struct promote<T, false> {};
+template <>
+struct promote<long double> {
+  using type = long double;
+};
+template <>
+struct promote<double> {
+  using type = double;
+};
+template <>
+struct promote<float> {
+  using type = float;
+};
+template <class T>
+using promote_t = typename promote<T>::type;
+template <class T, class U>
+struct promote_2 {
+  using type = decltype(promote_t<T>() + promote_t<U>());
+};
+template <class T, class U>
+using promote_2_t = typename promote_2<T, U>::type;
+}  // namespace Impl
+
 namespace Experimental {
 
 #if defined(KOKKOS_ENABLE_SYCL)
-#define NAMESPACE_MATH_FUNCTIONS sycl
+#define KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE sycl
+#else
+#define KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE std
+#endif
+
+// NOTE long double overloads are not available on the device
+#if defined(KOKKOS_ENABLE_CUDA) || defined(KOKKOS_ENABLE_HIP) || \
+    defined(KOKKOS_ENABLE_SYCL) || defined(KOKKOS_ENABLE_OPENMPTARGET)
 #else
-#define NAMESPACE_MATH_FUNCTIONS std
+#define KOKKOS_IMPL_MATH_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS
 #endif
 
-#define KOKKOS_IMPL_UNARY_FUNCTION_FLOATING_POINT(FUNC, RETURNTYPE, ARGTYPE) \
-  KOKKOS_INLINE_FUNCTION RETURNTYPE FUNC(ARGTYPE x) {                        \
-    using NAMESPACE_MATH_FUNCTIONS::FUNC;                                    \
-    return FUNC(x);                                                          \
+#if defined(KOKKOS_IMPL_MATH_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS)
```

Comment:
Why are there overloads other than for `long double` guarded here? I would prefer to avoid the duplicated code if possible.

Reply:
What alternative do you suggest?

Reply:
Something like
```diff
diff --git a/core/src/Kokkos_MathematicalFunctions.hpp b/core/src/Kokkos_MathematicalFunctions.hpp
index 50fde82d7..8c378559b 100644
--- a/core/src/Kokkos_MathematicalFunctions.hpp
+++ b/core/src/Kokkos_MathematicalFunctions.hpp
@@ -100,9 +100,7 @@ namespace Experimental {
 #define KOKKOS_IMPL_MATH_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS
 #endif
 
-#if defined(KOKKOS_IMPL_MATH_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS)
-
-#define KOKKOS_IMPL_MATH_UNARY_FUNCTION(FUNC)                                 \
+#define KOKKOS_IMPL_MATH_UNARY_FUNCTION_OTHER(FUNC)                           \
   KOKKOS_INLINE_FUNCTION float FUNC(float x) {                                \
     using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
     return FUNC(x);                                                           \
@@ -111,18 +109,10 @@ namespace Experimental {
     using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
     return FUNC(x);                                                           \
   }                                                                           \
-  KOKKOS_INLINE_FUNCTION long double FUNC(long double x) {                    \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
-    return FUNC(x);                                                           \
-  }                                                                           \
   KOKKOS_INLINE_FUNCTION float FUNC##f(float x) {                             \
     using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
     return FUNC(x);                                                           \
   }                                                                           \
-  KOKKOS_INLINE_FUNCTION long double FUNC##l(long double x) {                 \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
-    return FUNC(x);                                                           \
-  }                                                                           \
   template <class T>                                                          \
   KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_integral<T>::value, double> \
   FUNC(T x) {                                                                 \
@@ -130,7 +120,23 @@ namespace Experimental {
     return FUNC(static_cast<double>(x));                                      \
   }
 
-#define KOKKOS_IMPL_MATH_UNARY_PREDICATE(FUNC)                              \
+#if defined(KOKKOS_IMPL_MATH_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS)
+#define KOKKOS_IMPL_MATH_UNARY_FUNCTION(FUNC)                 \
+  KOKKOS_IMPL_MATH_UNARY_FUNCTION_OTHER(FUNC)                 \
+  KOKKOS_INLINE_FUNCTION long double FUNC(long double x) {    \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;         \
+    return FUNC(x);                                           \
+  }                                                           \
+  KOKKOS_INLINE_FUNCTION long double FUNC##l(long double x) { \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;         \
+    return FUNC(x);                                           \
+  }                                                           \
+#else
+#define KOKKOS_IMPL_MATH_UNARY_FUNCTION(FUNC) \
+  KOKKOS_IMPL_MATH_UNARY_FUNCTION_OTHER(FUNC)
+#endif
+
+#define KOKKOS_IMPL_MATH_UNARY_PREDICATE_OTHER(FUNC)                        \
   KOKKOS_INLINE_FUNCTION bool FUNC(float x) {                               \
     using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                       \
     return FUNC(x);                                                         \
@@ -139,10 +145,6 @@ namespace Experimental {
     using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                       \
     return FUNC(x);                                                         \
   }                                                                         \
-  KOKKOS_INLINE_FUNCTION bool FUNC(long double x) {                         \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                       \
-    return FUNC(x);                                                         \
-  }                                                                         \
   template <class T>                                                        \
   KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_integral<T>::value, bool> \
   FUNC(T x) {                                                               \
@@ -150,7 +152,19 @@ namespace Experimental {
     return FUNC(static_cast<double>(x));                                    \
   }
 
-#define KOKKOS_IMPL_MATH_BINARY_FUNCTION(FUNC)                               \
+#if defined(KOKKOS_IMPL_MATH_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS)
+#define KOKKOS_IMPL_MATH_UNARY_PREDICATE(FUNC)        \
+  KOKKOS_IMPL_MATH_UNARY_PREDICATE_OTHER(FUNC)        \
+  KOKKOS_INLINE_FUNCTION bool FUNC(long double x) {   \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC; \
+    return FUNC(x);                                   \
+  }
+#else
+#define KOKKOS_IMPL_MATH_UNARY_PREDICATE(FUNC) \
+  KOKKOS_IMPL_MATH_UNARY_PREDICATE_OTHER(FUNC)
+#endif
+
+#define KOKKOS_IMPL_MATH_BINARY_FUNCTION_OTHER(FUNC)                         \
   KOKKOS_INLINE_FUNCTION float FUNC(float x, float y) {                      \
     using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                        \
     return FUNC(x, y);                                                       \
@@ -181,57 +195,29 @@ namespace Experimental {
     return FUNC(static_cast<Promoted>(x), static_cast<Promoted>(y));         \
   }
 
-#else  // long double overloads are not available
-
-#define KOKKOS_IMPL_MATH_UNARY_FUNCTION(FUNC)                                 \
-  KOKKOS_INLINE_FUNCTION float FUNC(float x) {                                \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
-    return FUNC(x);                                                           \
-  }                                                                           \
-  KOKKOS_INLINE_FUNCTION double FUNC(double x) {                              \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
-    return FUNC(x);                                                           \
-  }                                                                           \
-  KOKKOS_INLINE_FUNCTION float FUNC##f(float x) {                             \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
-    return FUNC(x);                                                           \
-  }                                                                           \
-  template <class T>                                                          \
-  KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_integral<T>::value, double> \
-  FUNC(T x) {                                                                 \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                         \
-    return FUNC(static_cast<double>(x));                                      \
-  }
-
-#define KOKKOS_IMPL_MATH_UNARY_PREDICATE(FUNC)                              \
-  KOKKOS_INLINE_FUNCTION bool FUNC(float x) {                               \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                       \
-    return FUNC(x);                                                         \
-  }                                                                         \
-  KOKKOS_INLINE_FUNCTION bool FUNC(double x) {                              \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                       \
-    return FUNC(x);                                                         \
-  }                                                                         \
-  template <class T>                                                        \
-  KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_integral<T>::value, bool> \
-  FUNC(T x) {                                                               \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                       \
-    return FUNC(static_cast<double>(x));                                    \
+#if defined(KOKKOS_IMPL_MATH_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS)
+#define KOKKOS_IMPL_MATH_BINARY_FUNCTION(FUNC)                               \
+  KOKKOS_IMPL_MATH_BINARY_FUNCTION_OTHER(FUNC)                               \
+  KOKKOS_INLINE_FUNCTION long double FUNC(long double x, long double y) {    \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                        \
+    return FUNC(x, y);                                                       \
+  }                                                                          \
+  KOKKOS_INLINE_FUNCTION long double FUNC##l(long double x, long double y) { \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                        \
+    return FUNC(x, y);                                                       \
+  }                                                                          \
+  template <class T1, class T2>                                              \
+  KOKKOS_INLINE_FUNCTION std::enable_if_t<std::is_arithmetic<T1>::value &&   \
+                                              std::is_arithmetic<T2>::value, \
+                                          Kokkos::Impl::promote_2_t<T1, T2>> \
+  FUNC(T1 x, T2 y) {                                                         \
+    using Promoted = Kokkos::Impl::promote_2_t<T1, T2>;                      \
+    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                        \
+    return FUNC(static_cast<Promoted>(x), static_cast<Promoted>(y));         \
   }
-
+#else
 #define KOKKOS_IMPL_MATH_BINARY_FUNCTION(FUNC)                          \
-  KOKKOS_INLINE_FUNCTION float FUNC(float x, float y) {                 \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                   \
-    return FUNC(x, y);                                                  \
-  }                                                                     \
-  KOKKOS_INLINE_FUNCTION double FUNC(double x, double y) {              \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                   \
-    return FUNC(x, y);                                                  \
-  }                                                                     \
-  KOKKOS_INLINE_FUNCTION float FUNC##f(float x, float y) {              \
-    using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                   \
-    return FUNC(x, y);                                                  \
-  }                                                                     \
+  KOKKOS_IMPL_MATH_BINARY_FUNCTION_OTHER(FUNC)                          \
   template <class T1, class T2>                                         \
   KOKKOS_INLINE_FUNCTION std::enable_if_t<                              \
       std::is_arithmetic<T1>::value && std::is_arithmetic<T2>::value && \
@@ -243,7 +229,6 @@ namespace Experimental {
     using KOKKOS_IMPL_MATH_FUNCTIONS_NAMESPACE::FUNC;                   \
     return FUNC(static_cast<Promoted>(x), static_cast<Promoted>(y));    \
   }
-
 #endif
 
 // Basic operations
```
optionally combing all the branches.

Reply:
Nah I don't like it

Reply:
What's your argument for having the duplications then?

Reply:
It's what I was trying to move away from with the refactor.  The goal was less macro to reduce complexity and increase readability, even if it means duplicating code.

Reply:
I guess we disagree on this point here then. :slightly_smiling_face: